### PR TITLE
Implement parsing of override keyword

### PR DESCRIPTION
**Pull request by Marenz**
Doesn't break any compability yet, so still targeting develop at this point.
prepares for #5424

### CODE REVIEW

**Comment by chriseth** [comment]
Needs changes in ast export (and import...)

File: libsolidity/ast/AST.h [hunk 1]
```diff
@@ -624,6 +625,7 @@ class FunctionDefinition: public CallableDeclaration, public Documented, public
 		ImplementationOptional(_body != nullptr),
 		m_stateMutability(_stateMutability),
 		m_isConstructor(_isConstructor),
+		m_isOverriding(_isOverriding),
```
Comment (chriseth) [inline libsolidity/ast/AST.h]: maybe just `overriding`?

File: libsolidity/ast/AST.h [hunk 2]
```diff
@@ -753,6 +760,7 @@ class VariableDeclaration: public Declaration
 	bool m_isStateVariable; ///< Whether or not this is a contract state variable
 	bool m_isIndexed; ///< Whether this is an indexed variable (used by events).
 	bool m_isConstant; ///< Whether the variable is a compile-time constant.
+	bool m_isOverriding; ///< whether the variable is overriding a public base class variable
```
Comment (chriseth) [inline libsolidity/ast/AST.h]:
```suggestion
	bool m_isOverriding; ///< whether the variable is overriding an external base contract function
```

Comment (Marenz) [reply libsolidity/ast/AST.h]: A state variable can override a function fro the base contract?

**Comment by chriseth** [comment]
Looks good otherwise!

File: libsolidity/ast/AST.h (hunk 2 above)
Comment (chriseth) [reply libsolidity/ast/AST.h]: The only situation where `override` is meaningful for state variables is if the getter of the state variable overrides a function, and for that, the function in the base contract has to be external.

File: libsolidity/ast/AST.h (hunk 1 above)
Comment (Marenz) [reply libsolidity/ast/AST.h]: wouldn't that be inconsistent with the existing naming `isContructor` and `isIndexed` etc?

Comment (chriseth) [reply libsolidity/ast/AST.h]: actually the coding style says that `is` prefixes should be avoided, but yeah, we have a lot of them here.

**Comment by Marenz** [comment]
updated

**Comment by leonardoalt** [comment]
Tests failing, I think expectations need to be updated.

**Comment by Marenz** [comment]
fixed now. Had to investigate another bug that turned out to be just due to a missing `make clean`

**Comment by chriseth** [comment]
Are you planning to change this again to support the arguments?

**Comment by Marenz** [comment]
Yes, currently working on exactly that

File: libsolidity/parsing/Parser.cpp [hunk 3]
```diff
@@ -382,6 +382,34 @@ Declaration::Visibility Parser::parseVisibilitySpecifier()
...
+	ASTNodeFactory nodeFactory(*this);
+
+	m_scanner->next();
+
+	std::vector<ASTPointer<UserDefinedTypeName>> overrides;
+
+	if (m_scanner->currentToken() == Token::LParen)
+	{
+		m_scanner->next();
+		while (m_scanner->currentToken() == Token::Identifier)
```
Comment (ekpyron) [inline libsolidity/parsing/Parser.cpp]: What happens for let's say ``override(A, address)``?

Comment (chriseth) [reply libsolidity/parsing/Parser.cpp]: I think using `while (m_scanner->currentToken() != Token::RParen)` should fix that.

Comment (ekpyron) [reply libsolidity/parsing/Parser.cpp]: That might still fail (resp. "fail to fail") for ``override(A,)``, right? (which could be fixed by ``expectToken(Token::RParen)`` after the loop - by the way ``expectToken`` advances by default, doesn't it? Which means the ``m->scanner->next()`` after the ``expectToken(Token::Comma);`` might also have to be removed.)

Comment (ekpyron) [reply libsolidity/parsing/Parser.cpp]: Both should be test cases in any case.

Comment (ekpyron) [reply libsolidity/parsing/Parser.cpp]: Also ``override()`` should be a test case - I'd disallow that as well, but we could also say it's fine.

File: test/libsolidity/ASTJSON/override.json [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+						"statements" : []
+					},
+					"documentation" : null,
+					"id" : 29,
+					"implemented" : true,
+					"kind" : "function",
+					"modifiers" : [],
+					"name" : "faa",
+					"nodeType" : "FunctionDefinition",
+					"overrides" :
```
Comment (ekpyron) [inline test/libsolidity/ASTJSON/override.json:282]: Yep, this swallowed the ``B`` due to advancing after the ``expectToken`` above.

File: libsolidity/parsing/Parser.cpp (hunk 3 above)
Comment (chriseth) [reply libsolidity/parsing/Parser.cpp]: Oh of course, I did not mean that just modifying the while condition is the only required change.

**Comment by Marenz** [comment]
Updated. Should fix all your concerns :>

File: libsolidity/ast/AST.h [hunk 5]
```diff
@@ -604,6 +604,28 @@ class CallableDeclaration: public Declaration, public VariableScope
 	ASTPointer<ParameterList> m_returnParameters;
 };
```
Comment (chriseth) [inline libsolidity/ast/AST.h:606]: Please document.

File: libsolidity/ast/AST.h [hunk 6]
```diff
@@ -604,6 +604,28 @@ class CallableDeclaration: public Declaration, public VariableScope
...
+	):
+		ASTNode(_location),
+		m_overrides(_overrides)
+	{
+	}
+
+	void accept(ASTVisitor& _visitor) override;
+	void accept(ASTConstVisitor& _visitor) const override;
+
+	///< Returns the list of specific overrides, if any
```
Comment (chriseth) [inline libsolidity/ast/AST.h]:
```suggestion
	/// @returns the list of specific overrides, if any
```

File: libsolidity/ast/AST.h [hunk 7]
```diff
@@ -604,6 +604,28 @@ class CallableDeclaration: public Declaration, public VariableScope
...
+	std::vector<ASTPointer<UserDefinedTypeName>> overrides() const { return m_overrides; }
```
Comment (chriseth) [inline libsolidity/ast/AST.h]: Why not return by `const&`?

File: libsolidity/ast/AST.h [hunk 8]
```diff
@@ -613,6 +635,7 @@ class FunctionDefinition: public CallableDeclaration, public Documented, public
 		Declaration::Visibility _visibility,
 		StateMutability _stateMutability,
 		bool _isConstructor,
+		ASTPointer<OverrideSpecifier> const& _overrides,
```
Comment (chriseth) [inline libsolidity/ast/AST.h:642]: Does it make sense to move this all the way up to `Declaration`?

File: libsolidity/parsing/Parser.cpp [hunk 9]
```diff
@@ -382,6 +382,34 @@ Declaration::Visibility Parser::parseVisibilitySpecifier()
...
+			if (m_scanner->currentToken() == Token::RParen)
+				break;
+
+			expectToken(Token::Comma);
+		}
+
+		expectToken(Token::RParen);
+	}
+
+	return nodeFactory.createNode<OverrideSpecifier>(move(overrides));
```
Comment (chriseth) [inline libsolidity/parsing/Parser.cpp:412]: Please check that the sourcelocation end pointer is set properly here (in both cases)

File: libsolidity/parsing/Parser.cpp [hunk 10]
```diff
@@ -493,6 +522,13 @@ Parser::FunctionHeaderParserResult Parser::parseFunctionHeader(bool _forceEmptyN
 			else
 				result.stateMutability = parseStateMutability();
 		}
+		else if (token == Token::Override)
+		{
+			if (result.overrides)
+				parserError("Override already specified.");
+
```
Comment (chriseth) [inline libsolidity/parsing/Parser.cpp:531]: Maybe use `else`, since the user might not expect the override to be overridden?

Comment (chriseth) [reply libsolidity/parsing/Parser.cpp:531]: Ah ok sorry, this won't work because it would totally confuse the parser.

**Comment by chriseth** [comment]
Please also update grammar.txt

File: libsolidity/parsing/Parser.cpp [hunk 11]
```diff
@@ -493,6 +522,13 @@ Parser::FunctionHeaderParserResult Parser::parseFunctionHeader(bool _forceEmptyN
 			else
 				result.stateMutability = parseStateMutability();
 		}
+		else if (token == Token::Override)
```
Comment (chriseth) [inline libsolidity/parsing/Parser.cpp]: I think this should also take `allowModifiers &&`, at least `override` should not be allowed for function types, where `allowModifiers` is set to false. Maybe `allowModifiers` could be renamed.

File: libsolidity/parsing/Parser.cpp [hunk 12]
```diff
@@ -540,6 +576,7 @@ ASTPointer<ASTNode> Parser::parseFunctionDefinitionOrFunctionTypeStateVariable()
 			header.visibility,
```
Comment (chriseth) [inline libsolidity/parsing/Parser.cpp:579]: Could also add `header.overrides` to the condition in line 520/556, I think.

File: libsolidity/parsing/Parser.cpp (hunk 9 above)
Comment (Marenz) [reply libsolidity/parsing/Parser.cpp:412]: Check as in, with a test or simply manually checking?

Comment (Marenz) [reply libsolidity/parsing/Parser.cpp:412]: looking over the code again, I think they are fine. The c'tor sets the start position, the createNode call the end position

Comment (chriseth) [reply libsolidity/parsing/Parser.cpp:412]: Ok!

**Comment by Marenz** [comment]
> Could also add header.overrides to the condition in line 520/556, I think.

fixed all points but this one, wasn't sure how you meant this

File: libsolidity/ast/AST.h [hunk 13]
```diff
@@ -604,6 +604,32 @@ class CallableDeclaration: public Declaration, public VariableScope
 	ASTPointer<ParameterList> m_returnParameters;
 };
 
+/**
+ * Function override specifier. Consists of a single override keyword or a list of
```
Comment (chriseth) [inline libsolidity/ast/AST.h]:
```suggestion
 * Function override specifier. Consists of a single override keyword potentially followed by a parenthesized list of base contract names.
```

**Review by chriseth — APPROVED** [review]

**Comment by chriseth** [comment]
Hm, test seems to be failing, can you check, please?

**Comment by Marenz** [comment]
The failing test is the thing I mentioned in the meeting.. it seems the order of values in an json array is not consistent

**Comment by Marenz** [comment]
I'll take care of it

**Comment by Marenz** [comment]
This will now work once current develop is merged into develop_060

**Comment by Marenz** [comment]
I pushed some small fixes and rebased on latest develop_060. Here is the diff of my fixes only: https://github.com/ethereum/solidity/compare/aa0a53c0282391ae45dd7c118f918fea6ed72ac5..f4f0f3e84c2e2c02bb6319ab7418977260b5e3f6

**Comment by Marenz** [comment]
I investigated the doc failure. It's caused by a change @ekpyron sponsored. `solhint` runs on older grammar and thus will keep failing for this until we released 060 and they have updated their grammar.

So it's safe to ignore this for this PR.

File: libsolidity/parsing/Parser.cpp [hunk 14]
```diff
@@ -382,6 +382,36 @@ Declaration::Visibility Parser::parseVisibilitySpecifier()
...
 
+ASTPointer<OverrideSpecifier> Parser::parseOverrideSpecifier()
+{
+	solAssert(m_scanner->currentToken() == Token::Override, "");
+	ASTNodeFactory nodeFactory(*this);
+
+	nodeFactory.markEndPosition();
+	m_scanner->next();
+
+	std::vector<ASTPointer<UserDefinedTypeName>> overrides;
```
Comment (erak) [inline libsolidity/parsing/Parser.cpp]: Perhaps move this to line 389, such that all local variables are declared in the same location?

File: test/libsolidity/ASTJSON/override.json [hunk 15]
```diff
@@ -0,0 +1,337 @@
...
+						"statements" : []
+					},
+					"documentation" : null,
+					"id" : 15,
+					"implemented" : true,
+					"kind" : "function",
+					"modifiers" : [],
+					"name" : "faa",
+					"nodeType" : "FunctionDefinition",
+					"overrides" : [],
```
Comment (erak) [inline test/libsolidity/ASTJSON/override.json:163]: What's the difference between `"overrides" : []` and `"overrides" : null` ?

File: test/libsolidity/ASTJSON/non_utf8.sol [hunk 16]
```diff
@@ -1 +1,3 @@
 contract C { function f() public { var x = hex"ff"; } }
+
+// ----
```
Comment (erak) [inline test/libsolidity/ASTJSON/non_utf8.sol:3]: Are those delimiters also been made obligatory in ASTJSON tests with this PR?

File: test/libsolidity/ASTJSON/override.json (hunk 15 above)
Comment (Marenz) [reply test/libsolidity/ASTJSON/override.json:163]: I believe `[]` means that `override` was specified but no list of things overriden was. E.g.

```
function foo() public overrides {}  // overrides: []
function foo() public overrides(A, B) {} // overrides: ["A", "B"]
function foo() public {} // overrides: null
```

File: test/libsolidity/ASTJSON/non_utf8.sol (hunk 16 above)
Comment (Marenz) [reply test/libsolidity/ASTJSON/non_utf8.sol:3]: No, I think they were just added by the automatic expectation updater

File: libsolidity/parsing/Parser.cpp (hunk 14 above)
Comment (Marenz) [reply libsolidity/parsing/Parser.cpp]: Moved it up.

File: test/libsolidity/ASTJSON/non_utf8.sol (hunk 16 above)
Comment (erak) [reply test/libsolidity/ASTJSON/non_utf8.sol:3]: Ok!

File: test/libsolidity/ASTJSON/override.json (hunk 15 above)
Comment (erak) [reply test/libsolidity/ASTJSON/override.json:163]: Makes sense, thanks!

**Review by erak — APPROVED** [review]
Looks good! Need to track the removal of `chk_docs_examples` from Circle for `develop_060` as suggested by @ekpyron though.

**Comment by Marenz** [comment]
I forgot to add one test case that is now added.

**Pull request merged**